components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -1283,24 +1293,27 @@ export class GitpodServerImpl<Client extends GitpodClient, Server extends Gitpod
             }
         }
 
-        const existingVar = existingVars.find(v => v.name == variable.name && v.repositoryPattern == variable.repositoryPattern);
-        if (!!existingVar) {
-            // overwrite existing variable rather than introduce a duplicate
-            variable.id = existingVar.id;
-        }
-
         const envvar: UserEnvVar = {
             ...variable,
             id: variable.id || uuidv4(),
             userId: user.id,
         };
+        await this.guardAccess({ kind: 'envVar', subject: envvar }, typeof variable.id === 'string' ? 'update' : 'create');
+
         await this.userDB.setEnvVar(envvar);
     }
 
     async deleteEnvVar(variable: UserEnvVarValue): Promise<void> {
         // Note: this operation is per-user only, hence needs no resource guard
         const user = this.checkUser("deleteEnvVar");
 
+        if (!variable.id && variable.name && variable.repositoryPattern) {
```

Comment:
This seems to be the same/a similar check as above. Would it make sense to extract those into one method `isUpdate`?

Reply:
I would prefer to keep code like that, since I'm not sure about client expectation when `id` is provided for delete.

Ideally I would remove `id` completely and use `(name, repositoryPattern)` as a primary key.